[FLINK-4067] [runtime] Add savepoint headers

uce opened the pull request:
Savepoints were previously persisted without any meta data using default Java serialization of a `CompletedCheckpoint`. This PR introduces a savepoint class with version-specific serializers and stores savepoints with meta data.

Savepoints expose a version number and a `Collection<TaskState>` for savepoint restore.

Currently, there ~~are two savepoint versions~~ is only one savepoint version:

~~- `SavepointV0` (Flink 1.0): This is a wrapper around the Flink 1.0 `CompletedCheckpoint`. Because this class changed since Flink 1.0 I had to re-introduce it with this PR in order to allow restoring from Flink 1.0 savepoints (otherwise the old `StateForTask` state can not be deserialized via default Java serialization). Therefore, I've renamed the Flink 1.1 `CompletedCheckpoint` to `Checkpoint`. I've annotated the re-introduced classes with `@Deprecated`. We can remove those with Flink 1.2.~~
- `SavepointV0` (Flink 1.1): This is the current savepoint version, which holds a reference to the `Checkpoint` task state collection, but is serialized with a custom serializater not relying on default Java serialization. Therefore, it should not happen again that we need to stick to certain classes in future Flink versions.

The savepoints are stored in `FsSavepointStore` with the following format:

```
MagicNumber SavepointVersion Savepoint
  - MagicNumber => int
  - SavepointVersion => int (returned by Savepoint#getVersion())
  - Savepoint => bytes (serialized via version-specific SavepointSerializer)
```

The header is minimal (magic number, version). All savepoint-specific meta data can be moved to the savepoint itself. This is also were we would have to add new meta data in future versions, allowing us to differentiate between different savepoint versions when we change the serialization stack etc.

~~I've tested both manually and via a unit test that it works to trigger a savepoint via Flink 1.0 and resume from it with Flink 1.1.~~ It's not possible to resume from a Flink 1.0 savepoint.

All savepoint related classes have been moved from `checkpoint` to a new sub package `checkpoint.savepoint`.

The main classes to look at to review this PR are `Savepoint` and subclasses, `SavepointSerializer` and subclasses, and `FsSavepointStore`. @tillrohrmann do you have time to do this?

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/Checkpoint.java:
```diff
@@ -0,0 +1,135 @@
+		long completionTimestamp,
+		Map<JobVertexID, TaskState> taskStates) {
+
+		this.job = job;
```
`checkNotNull`?

uce replied:
Will do (this is an unrelated change though as `Checkpoint` is a renaming of the current `master`'s `CompletedCheckpoint`)

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointRecoveryFactory.java:
```diff
@@ -41,13 +41,13 @@
+	 * @return {@link CheckpointStore} instance for the job
 	 */
-	CompletedCheckpointStore createCompletedCheckpoints(JobID jobId, ClassLoader userClassLoader)
+	CheckpointStore createCompletedCheckpoints(JobID jobId, ClassLoader userClassLoader)
```
Maybe we could rename this method to `createCheckpointStore` if not public.

uce replied:
Will do, the current name is not good

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/PendingCheckpoint.java:
```diff
@@ -109,13 +109,13 @@ public boolean isDiscarded() {
 	}
 	
-	public CompletedCheckpoint toCompletedCheckpoint() {
+	public Checkpoint toCompletedCheckpoint() {
```
Maybe we could rename this method to something `finalizeCheckpoint` or `completeCheckpoint` if it's not public.

uce replied:
Will do

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/StandaloneCheckpointRecoveryFactory.java:
```diff
@@ -37,10 +37,10 @@ public void stop() {
 
 	@Override
-	public CompletedCheckpointStore createCompletedCheckpoints(JobID jobId, ClassLoader userClassLoader)
+	public CheckpointStore createCompletedCheckpoints(JobID jobId, ClassLoader userClassLoader)
```
Renaming?

uce replied:
Yes, to be consistent with the `CompletedCheckpoint => Checkpoint` renaming

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/savepoint/FsSavepointStore.java:
```diff
@@ -0,0 +1,225 @@
+		FSDataOutputStream fdos = null;
+
+		// Try to create a FS output stream
+		for (int attempt = 0; attempt < 10; attempt++) {
```
What's usually the reason why `fileSystem.create` can fail? Why did you choose 9 retries?

uce replied:
Since it can be also a distributed FS like HDFS, it can fail for various reasons. The code here is moved from what we currently have in master (including number of retries etc.). I think initially from `FileSystemStateStorageHelper` to `FileSystemStateStore` to here.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/savepoint/FsSavepointStore.java:
```diff
@@ -0,0 +1,225 @@
+
+		try {
+			try (FSDataInputStream fdis = createFsInputStream(new Path(path))) {
+				try (DataInputStream dis = new DataInputStream(fdis)) {
```
Why do we have a triple nested try block here? Wouldn't it also work to write:

```
try (FSDataInputStream fdis = createFsInputStream(new Path(path));
    DataInputStream dis = new DataInputStream(fdis)) {
    ...
} catch (Throwable t) {
...
}
```

uce replied:
True :smile:

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/savepoint/FsSavepointStore.java:
```diff
@@ -0,0 +1,225 @@
+					}
+				}
+			}
+		} catch (Throwable t) {
```
Exceptions which are thrown due to a wrong path or because the input stream couldn't be opened will also trigger the savepoint version 0 loading. Maybe we could differentiate between unable to open file stream and wrong magic number or unable to read version.

uce replied:
Good point, especially with opening the stream etc.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/savepoint/SavepointStore.java:
```diff
@@ -0,0 +1,71 @@
+	 *
+	 * @throws Exception Failures during shut down are forwarded
+	 */
+	void shutDown() throws Exception;
```
I think so far we've usually written shutdown without a capital letter.

uce replied:
We have it as `shutDown` in other places as well, but it's true that there are many places with `shutdown`. Since the method name describes an action I would like to keep it as two words *shut down*.

tillrohrmann replied:
A quick search in the code base revealed that we have 5 usages of `shutDown()` and 100+ usages of `shutdown()`. I don't want to be hairsplitting, but I think the current convention is to write it without a capital letter.

uce replied:
But the convention for virtually all other methods is that method names are typically verbs or verb phrases. But I don't want to be hairsplitting. I'll rename it.

tillrohrmann replied:
I think *shutdown* is a verb in fact.

uce replied:
http://www.dict.cc/?s=shut+down vs. http://www.dict.cc/?s=shutdown, but I really don't care :smile: Let's just leave it as `shutdown`.

tillrohrmann replied:
Hmm dict.leo.org says it differently: http://dict.leo.org/ende/index_de.html#/search=shutdown&searchLoc=0&resultOrder=basic&multiwordShowSingle=on&pos=0

But on the other hand there is also this site: http://notaverb.com/shutdown.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/savepoint/FsSavepointStore.java:
```diff
@@ -0,0 +1,225 @@
+	/**
+	 * Tries to load a Flink 1.0 savepoint, which was not stored with a header.
+	 *
+	 * <p>This can be removed for Flink 1.2.
```
Why can we remove this for Flink 1.2? Wouldn't that mean that Flink 1.2 can only restore savepoints from version 1.1 onwards? I think it should also be possible to restore savepoints from version 1.0 with version 1.2.

uce replied:
True, that should be defined outside of the code and we will probably like to keep it backwards compatible as long as possible. I will remove all comments like this.

StephanEwen commented:
test comment

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/savepoint/SavepointV0Serializer.java:
```diff
@@ -0,0 +1,58 @@
+ */
+public class SavepointV0Serializer implements SavepointSerializer<SavepointV0> {
+
+	public static final SavepointV0Serializer INSTANCE = new SavepointV0Serializer();
```
Maybe we should add a private constructor so that nobody can instantiate this class.

uce replied:
Forgot that, will add the private constructor

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/savepoint/SavepointV0Serializer.java:
```diff
@@ -0,0 +1,58 @@
+	@Override
+	public SavepointV0 deserialize(InputStream is) throws IOException {
+		try (ObjectInputStream ois = new ObjectInputStream(is)) {
+			try {
```
Isn't one try block enough?

uce replied:
Oh yes 😄

tillrohrmann commented on flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/savepoint/FsSavepointStoreTest.java:
```diff
@@ -0,0 +1,308 @@
+		String pathNewSavepoint = store.storeSavepoint(newSavepoint);
+		assertEquals(1, tmp.getRoot().listFiles().length);
+
+		// Savepoint v0
```
savepoint v1

uce replied:
True

tillrohrmann commented on flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/savepoint/FsSavepointStoreTest.java:
```diff
@@ -0,0 +1,308 @@
+	}
+
+	/**
+	 * Tests that an exception during store cleans up any created files.
```
Shouldn't we check that only the failed savepoint is cleaned up?

uce replied:
Misleading formulation in the comment, that's exactly what's tested.

tillrohrmann commented:
Really good work @uce. The code is well structured and thoroughly tested. I had only some minor comments.

While testing the code with the streaming state machine job I stumbled across a problem, though. Recovering from a Flink 1.0 savepoint does not work if the job contains a `keyBy` operation. The reason is that we had a faulty murmur hash implementation in Flink 1.0 and due to its correction, the mapping of keys to sub tasks has changed. Consequently, the restored state no longer matches the assigned key spaces for each operator. This is the problematic [commit](https://github.com/apache/flink/commit/641a0d436c9b7a34ff33ceb370cf29962cac4dee).

Thus, this change is actually breaking our backwards compatibility with respect to savepoints. In order to solve the problem I see three possibilities:
- Revert the changes of this commit. But we don't know how the flawed murmur hash performs.
- Develop a tool which can repartition savepoints
- Don't support backwards compatibility between version 1.0 and 1.1

I think that option 3 is not doable given our backwards compatibility promise. Furthermore, option 2 is not really straight forward, if the user has a keyed stream where he uses the `Checkpointed` interface. Given that the release is upcoming, I think option 1 would be the best way to solve the problem.

uce commented:
Thanks for the thorough review! I didn't catch that by my manual test. I would also be in favour of 1) for the same reasons.

uce commented:
@greghogan You fixed the murmur hash in the mentioned commit. Do you think it's OK to revert back to the wrong version we had before? Did you actually observe bad behaviour of the wrong murmur hash variant?

uce commented:
Thanks for the review again. I've addressed all comments. If you have time, it would be great to have a look over the change in 157c4e1. I've ran all tests locally after reverting FLINK-3623 and everything passes.

uce commented:
If there are no objections, I would like to merge this.

uce commented:
After talking with @tillrohrmann and @StephanEwen, we discovered some obvious cases where restoring from 1.0 savepoints won't work in 1.1, for example if window state is involved, which is serialized in a different format since 1.0. I think the same is true for RocksDB.

Should we by default throw an Exception when a user tries to restore a Flink 1.0 savepoint with 1.1? On the other hand, there  are cases where it would work as expected... but if it does not work the error message will not be very helpful.

uce commented:
Magic number mismatch will fail with an error message stating that a likely cause for this is resuming from Flink 1.0 savepoints, which is not supported. I've removed the legacy classes and undid the renamings. `SavepointV1` (Flink 1.1) has been down-graded to `SavepointV0` and the previous `SavepointV0` (for Flink 1.0) has been removed.
